### PR TITLE
Pre-Dafny 4 edits to the reference manual

### PR DESCRIPTION
**Pull request by davidcok**
Part of #3397

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: docs/DafnyRef/Introduction.md [hunk 1]
```diff
@@ -60,7 +58,23 @@ initial section that deals with the lowest level constructs.
...
+## 1.1. Dafny 4.0
+
+The most recent major version of the Dafny language is Dafny 4.0, released in February 2023.
+It has some backwards incompatibilities with Dafny 3, as decribed in the [migration guide](https://github.com/dafny-lang/ide-vscode/wiki/Quick-migration-guide-from-Dafny-3.X-to-Dafny-4.0).
+
+Some of the highlights of Dafny 4.0 are
+* support for full unicode
+TODO
+
+The user documentation has been expanded with more examples, A FAQ, and an error explanation catalog.
```
Comment (stefan-aws) [inline docs/DafnyRef/Introduction.md:70]:
```suggestion
The user documentation has been expanded with more examples, a FAQ, and an error explanation catalog.
```

File: docs/DafnyRef/Introduction.md [hunk 2]
```diff
@@ -60,7 +58,23 @@ initial section that deals with the lowest level constructs.
...
+There is even a new book, [Program Proofs](https://mitpress.mit.edu/9780262546232/program-proofs/) by Dafny designer Rustan Leino.
+
+The IDE now has a framework for showing error explanation informatino and corresponding quick fixes are
```
Comment (stefan-aws) [inline docs/DafnyRef/Introduction.md:73]:
```suggestion
The IDE now has a framework for showing error explanation information and corresponding quick fixes are
```

File: docs/DafnyRef/Modules.md [hunk 3]
```diff
@@ -275,14 +275,14 @@ module Helpers {
   }
 }
 module Mod {
-  import opened Helpers
+  import opened H = Helpers
   function addOne(n: nat): nat {
     n - 1
```
Comment (stefan-aws) [inline docs/DafnyRef/Modules.md:280]:
```suggestion
    n + 1
```

File: docs/DafnyRef/Modules.md [hunk 4]
```diff
@@ -275,14 +275,14 @@ module Helpers {
   }
 }
 module Mod {
-  import opened Helpers
+  import opened H = Helpers
   function addOne(n: nat): nat {
```
Comment (stefan-aws) [inline docs/DafnyRef/Modules.md:279]:
```suggestion
  function addOne(n: nat): nat {
  // incorrect implementation
```

File: docs/DafnyRef/Modules.md [hunk 5]
```diff
@@ -340,7 +340,8 @@ module X {
 }
 ```
 `Option.a` now means the `a` in the datatype instead of the `a` in the module.
-To avoid confusion in such cases, it is an ambiguity error to `import open`
+To avoid confusion in such cases, it is an ambiguity error to use such a name
```
Comment (stefan-aws) [inline docs/DafnyRef/Modules.md:343]: I find this sentence a bit hard to parse.

File: docs/DafnyRef/Programs.md [hunk 6]
```diff
@@ -4,16 +4,16 @@ At the top level, a Dafny program (stored as files with extension `.dfy`)
...
 constants, methods, functions, lemmas, types (classes, traits, inductive and
 coinductive datatypes, newtypes, type synonyms, opaque types, and
-iterators) and modules, where the order of introduction is irrelevant. A
-class also contains a set of declarations, introducing fields, methods,
+iterators) and modules, where the order of introduction is irrelevant. 
+Some types, notably classes, also may contain a set of declarations, introducing fields, methods,
 and functions.
 
 When asked to compile a program, Dafny looks for the existence of a
 Main() method. If a legal Main() method is found, the compiler will emit
```
Comment (stefan-aws) [inline docs/DafnyRef/Programs.md:12]:
```suggestion
`Main()` method. If a legal `Main()` method is found, the compiler will emit
```

File: docs/DafnyRef/Programs.md [hunk 7]
```diff
@@ -4,16 +4,16 @@ At the top level, a Dafny program (stored as files with extension `.dfy`)
...
 an executable appropriate to the target language; otherwise it will emit
 a library or individual files.
 The conditions for a legal Main() method are described in the User Guide
```
Comment (stefan-aws) [inline docs/DafnyRef/Programs.md:15]:
```suggestion
The conditions for a legal `Main()` method are described in the User Guide
```

File: docs/DafnyRef/Programs.md [hunk 8]
```diff
@@ -4,16 +4,16 @@ At the top level, a Dafny program (stored as files with extension `.dfy`)
...
-([Section 13.7.1](#sec-user-guide-main)).
+([Section 3.4](#sec-user-guide-main)).
 If there is more than one Main(), Dafny will emit an error message.
```
Comment (stefan-aws) [inline docs/DafnyRef/Programs.md:17]:
```suggestion
If there is more than one `Main()`, Dafny will emit an error message.
```

File: docs/DafnyRef/Programs.md [hunk 9]
```diff
@@ -40,27 +40,33 @@ a file that will be included in the Dafny source. These included
...
+specified.
 
-The file name may be a path using the customary `/`, `.`, and `..` specifiers.
+The file name may be a path using the customary `/`, `.`, and `..` punctuation.
 The interpretation of the name (e.g., case-sensitivity) will depend on the
 underlying operating system. A path not beginning with `/` is looked up in
-the underlying file system relative to the current working directory (the
-one in which the dafny tool is invoked). Paths beginning with a device
+the underlying file system relative to the _location of the file in which the
+include directive is stated_. Paths beginning with a device
```
Comment (stefan-aws) [inline docs/DafnyRef/Programs.md:50]:
```suggestion
include directive is stated. Paths beginning with a device
```

File: docs/DafnyRef/Programs.md [hunk 10]
```diff
@@ -40,27 +40,33 @@ a file that will be included in the Dafny source. These included
...
+explicitly on the command line or the `--verify-included-files` option is
+specified.
 
-The file name may be a path using the customary `/`, `.`, and `..` specifiers.
+The file name may be a path using the customary `/`, `.`, and `..` punctuation.
 The interpretation of the name (e.g., case-sensitivity) will depend on the
 underlying operating system. A path not beginning with `/` is looked up in
-the underlying file system relative to the current working directory (the
-one in which the dafny tool is invoked). Paths beginning with a device
+the underlying file system relative to the _location of the file in which the
```
Comment (stefan-aws) [inline docs/DafnyRef/Programs.md:49]:
```suggestion
the underlying file system relative to the location of the file in which the
```

File: docs/DafnyRef/Programs.md [hunk 11]
```diff
@@ -135,13 +141,67 @@ implicitly ghost (non-ghost).
...
+may also be executable.
+
+To be executable, the program must have exactly one `Main` method and that 
+method must be a legal main entry point.
+
+* The program is searched for a method with the attribute `{:main}`.
+If exactly one is found, that method is used as the entry point; if more
+than one method has the `{:main}` attribute, an error message is issued.
+* Otherwise, the program is searched for a method with the name `Main`.
+If more than one is found
```
Comment (stefan-aws) [inline docs/DafnyRef/Programs.md:166]: Remove the line break here?

Comment (stefan-aws) [reply docs/DafnyRef/Programs.md:166]: Same below?

File: docs/DafnyRef/Programs.md [hunk 12]
```diff
@@ -135,13 +141,67 @@ implicitly ghost (non-ghost).
...
+  then the enclosing type must, when instantiated with auto-initializing
+  type parameters, be an auto-initializing type.
+  In this case, the runtime system will
+  invoke the entry-point method on a value of the enclosing type.
+
+Note, however, that the following are allowed:
+
+* The method is allowed to have `ensures` clauses
+* The method is allowed to have `decreases` clauses, including a
+  `decreases *`. (If Main() has a `decreases *`, then its execution may
```
Comment (stefan-aws) [inline docs/DafnyRef/Programs.md:194]:
```suggestion
  `decreases *`. (If `Main()` has a `decreases *`, then its execution may
```

Comment (stefan-aws) [reply docs/DafnyRef/Programs.md:194]: Sometimes you write `Main()` and sometimes `Main`, without brackets.

**Review by stefan-aws — CHANGES_REQUESTED** [review]
Thank you, looks good. Found a few small typos.

File: docs/DafnyRef/Introduction.md [hunk 13]
```diff
@@ -60,7 +58,23 @@ initial section that deals with the lowest level constructs.
...
 
-## 1.1. Dafny Example {#sec-example}
+## 1.1. Dafny 4.0
+
+The most recent major version of the Dafny language is Dafny 4.0, released in February 2023.
+It has some backwards incompatibilities with Dafny 3, as decribed in the [migration guide](https://github.com/dafny-lang/ide-vscode/wiki/Quick-migration-guide-from-Dafny-3.X-to-Dafny-4.0).
+
+Some of the highlights of Dafny 4.0 are
+* support for full unicode
+TODO
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/DafnyRef/Introduction.md:68]: Thanks for the reminder -- I do need input from Robin and others to fill this section out, but that can be a later PR.

File: docs/DafnyRef/Introduction.md (hunk 1 above)
Comment (davidcok) [reply docs/DafnyRef/Introduction.md:70]: I would say 'a FAQ' because I pronounce it as an acronym, though I understand the pronunciation varies. Even then 'an FAQ' is worse than 'a F-A-Q'. Ii should not be uppercase though

File: docs/DafnyRef/Modules.md (hunk 4 above)
Comment (davidcok) [reply docs/DafnyRef/Modules.md:279]: The function is not incorrect, though may be in practice it would be confusing.

File: docs/DafnyRef/Modules.md (hunk 5 above)
Comment (davidcok) [reply docs/DafnyRef/Modules.md:343]: It is a complex corner case. I revised.

File: docs/DafnyRef/Programs.md (hunk 9 above)
Comment (davidcok) [reply docs/DafnyRef/Programs.md:50]: The _ is the end of an emphasis. Changed line breaks to make it clearer.

File: docs/DafnyRef/Programs.md (hunk 10 above)
Comment (davidcok) [reply docs/DafnyRef/Programs.md:49]: Changed line breaks to make it clearer that this is the beginning of an emphasis

File: docs/DafnyRef/Programs.md (hunk 11 above)
Comment (davidcok) [reply docs/DafnyRef/Programs.md:166]: Done.

File: docs/DafnyRef/Programs.md (hunk 9 above)
Comment (stefan-aws) [reply docs/DafnyRef/Programs.md]: Oh, I see it now.

File: docs/DafnyRef/Programs.md [hunk 14]
```diff
@@ -191,7 +191,7 @@ Note, however, that the following are allowed:
 
 * The method is allowed to have `ensures` clauses
 * The method is allowed to have `decreases` clauses, including a
-  `decreases *`. (If Main() has a `decreases *`, then its execution may
+  `decreases *`. (If `Main()` has a `decreases *`, then its execution may
   go on forever, but in the absence of a `decreases *` on Main(), `dafny`
```
Comment (stefan-aws) [inline docs/DafnyRef/Programs.md:195]:
```suggestion
  go on forever, but in the absence of a `decreases *` on `Main()`, `dafny`
```

**Review by stefan-aws — APPROVED** [review]

**Review by stefan-aws — APPROVED** [review]
